src/base/tool-manager/tool-manager.ts

```diff
@@ -0,0 +1,143 @@
+/*******************************************************************************
+ * Copyright (c) 2018 EclipseSource
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * 	Philip Langer - initial API and implementation
+ ******************************************************************************/
+import { inject, injectable } from "inversify";
+import { TYPES } from "../types";
+import { EnableStandardToolsAction, EnableToolsAction, Tool } from "./tool";
+import { IActionHandlerInitializer, IActionHandler, ActionHandlerRegistry } from "../actions/action-handler";
+import { Action } from "../actions/action";
+import { ICommand } from "../commands/command";
+import { KeyListener } from "../views/key-tool";
+import { SModelElement } from "../model/smodel";
+import { matchesKeystroke } from "../../utils/keyboard";
+
+/**
+ * A tool manager coordinates the state of tools in the context of an editor.
+ *
+ * One instance of a tool manager is intended per editor, coordinating the state of all tools within
+ * this editor. A tool can be active or not. A tool manager ensures that activating a set of tools
+ * will disable all other tools, allowing them to invoke behavior when they become enabled or disabled.
+ */
+export interface ToolManager {
+
+    /** All tools managed by this tool manager. */
+    readonly managedTools: Tool[];
+
+    /** The tools that are enabled by standard, whenever no other tool is enabled. */
+    readonly standardTools: Tool[];
+
+    /** The currently active tools, which are either specifically enabled tools, or the standard tools. */
+    readonly activeTools: Tool[];
+
+    /**
+     * Enables the tools with the specified `toolIds`.
+     * Therefore, this manager first disables currently active tools and then enable the
+     * tools indicated in `toolIds`, making them the currently active tools. If this manager
+     * doesn't manage one or more tools specified in `toolIds`, it'll do nothing. If not a
+     * single tool that shall be enabled was found in the managed tools, it'll fall back to
+     * the standard tools.
+     *
+     * @param tools The tools to be enabled.
+     */
+    enable(toolIds: string[]): void;
+
+    /**
+     * Enables all standard tools.
+     */
+    enableStandardTools(): void;
+
+    /** Disables all currently active tools. After this call, no tool will be active anymore. */
+    disableActiveTools(): void;
+
+    registerStandardTools(...tools: Tool[]): void;
+
+    registerTools(...tools: Tool[]): void;
+}
+
+@injectable()
+export class DefaultToolManager implements ToolManager {
+
+    readonly standardTools: Tool[] = [];
+    readonly tools: Tool[] = [];
+    private actives: Tool[] = [];
```

Comment:
Ok. I didn't use readonly because of line 81 where we do `this.actives = []`. The available methods on arrays are a bit limited (e.g. there is no `clear()`). But in a subsequent commit, I made actives readonly and splice the array to clear it.